### PR TITLE
Sig windows tests

### PR DESCRIPTION
**Pull request by vyta**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Adding a passing density test similar to the one in e2e-node, but for windows nodes. This will help get the ball rolling for writing more windows specific tests

**Which issue(s) this PR fixes**:

Could help with windows related issues.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
/sig windows

### CODE REVIEW

File: test/e2e/windows/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Notes to run sig-windows tests
```
Comment (fejta) [inline test/e2e/windows/README.md]: Please create an OWNERS file in this new windows dir

File: test/e2e/windows/framework.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (yujuhong) [inline test/e2e/windows/framework.go]: 2018

File: test/e2e/windows/density.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+	stats "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	"k8s.io/apimachinery/pkg/util/uuid"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = SIGDescribe("Density [Serial] [Slow]", func() {
```
Comment (yujuhong) [inline test/e2e/windows/density.go]: I think this is likely to be added to a different suite, e.g., add a `[Feature:PerformanceBenchmark]` or something else.

File: test/e2e/windows/density.go [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+	Context("create a batch of pods", func() {
+		// TODO(coufon): the values are generous, set more precise limits with benchmark data
+		// and add more tests
+		dTests := []densityTest{
+			{
+				podsNr:   10,
+				interval: 0 * time.Millisecond,
+				cpuLimits: framework.ContainersCPUSummary{
+					stats.SystemContainerKubelet: {0.50: 0.30, 0.95: 0.50},
+					stats.SystemContainerRuntime: {0.50: 0.40, 0.95: 0.60},
```
Comment (yujuhong) [inline test/e2e/windows/density.go]: I don't think system containers are supported on windows. Remove these?

File: test/e2e/windows/density.go [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+				},
+				memLimits: framework.ResourceUsagePerContainer{
```
Comment (yujuhong) [inline test/e2e/windows/density.go]: Same here. There is no resource usage for kubelet/runtime on windows IIUC.

File: test/e2e/windows/density.go [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+	go controller.Run(stopCh)
+	defer close(stopCh)
+
+	// TODO(coufon): in the test we found kubelet starts while it is busy on something, as a result 'syncLoop'
+	// does not response to pod creation immediately. Creating the first pod has a delay around 5s.
+	// The node status has already been 'ready' so `wait and check node being ready does not help here.
+	// Now wait here for a grace period to let 'syncLoop' be ready
+	time.Sleep(sleepBeforeCreatePods)
+
+	//rc.Start()
```
Comment (yujuhong) [inline test/e2e/windows/density.go]: remove this

File: test/e2e/windows/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+1. Prereqs:
+    * KUBECONFIG=path/to/kubeconfig
+    * curl https://raw.githubusercontent.com/e2e-win/e2e-win-prow-deployment/master/repo-list.txt -o repo_list.yaml
```
Comment (yujuhong) [inline test/e2e/windows/README.md]: This should be updated to using the official sig-windows testing repo soon.

File: test/e2e/windows/density.go [hunk 8]
```diff
@@ -0,0 +1,319 @@
+/*
```
Comment (yujuhong) [inline test/e2e/windows/density.go]: Could you summarize what has changed (if there is any) from `e2e_node/density_test.go`?

**Review by yujuhong — COMMENTED** [review]
In general, we may want to add a `[Feature: Windows]` tag for all windows-only tests. Otherwise, one will have to go change all test configurations to exclude windows tests specifically.

Is there any plan to add a test job to run the density tests?

File: test/e2e/windows/README.md (hunk 7 above)
Comment (vyta) [reply test/e2e/windows/README.md]: Thanks for the review @yujuhong. Which repo are you referring to?

**Comment by PatrickLang** [comment]
@yujuhong I think we'll need to add a new job with either focus=[feature:windows] or [sig-windows]. We will need to write a lot of tests that will be similar to conformance tests that we need to include as well.

**Comment by fejta** [comment]
Please pass tests

**Comment by PatrickLang** [comment]
@vyta here's a suggested OWNERS file https://github.com/PatrickLang/kubernetes/commit/dccfe1c43ade4e49ca0e94eb9856817bed5289df - can you add it to this PR?

**Comment by PatrickLang** [comment]
https://github.com/PatrickLang/kubernetes/commit/18e89c7e7fee069a6ffa2101d0a873c238dc7715 might fix the bazel build but I haven't confirmed yet

**Comment by PatrickLang** [comment]
Still trying to get Bazel working, will continue tomorrow. @vyta here's my notes on it so far https://gist.github.com/PatrickLang/31b974bb576ef7767bbbf73f46cab382

**Comment by PatrickLang** [comment]
@vyta `hack/update-bazel.sh` threw an error for me, but it still fixed up the BUILD files. `make bazel-build` succeeds now with https://github.com/PatrickLang/kubernetes/commit/a4af001baeb88ed1830d66e80e71a566190db0f3 added

**Comment by adelina-t** [comment]
cc: @bclau

**Comment by vyta** [comment]
@PatrickLang thanks for looking into the bazel build

**Comment by bclau** [comment]
It seems that there are still a couple of issue with this PR. You could take a look at my commit, which also adds the test/e2e/windows folder and I've already taken care of these issues. Also, for the tests I've added, I've also added a skip check for non-windows OSes, so these tests won't run on Linux.

[1] https://github.com/kubernetes/kubernetes/pull/71468/commits/aac1d70a3bd5e0f028daf2a57e81d49701e43fad
[2] https://github.com/kubernetes/kubernetes/pull/71468/commits/aac1d70a3bd5e0f028daf2a57e81d49701e43fad#diff-574df117b4782ee7ccb3b6ee22ef72edR32

**Comment by vyta** [comment]
@bclau thanks! I'll make a couple changes to this PR based on your commit.

File: test/e2e/windows/README.md (hunk 7 above)
Comment (yujuhong) [reply test/e2e/windows/README.md]: This one, and I filed an issue: https://github.com/kubernetes-sigs/windows-testing/issues/1

@PatrickLang @adelina-t could we move the shared e2e utilities to the `windows-testing` repo, while waiting for the e2e test tagging to mature?

**Comment by yujuhong** [comment]
> @yujuhong I think we'll need to add a new job with either focus=[feature:windows] or [sig-windows]. We will need to write a lot of tests that will be similar to conformance tests that we need to include as well.

Either would work. I don't have a strong opinion on this. The downside of using [sig-windows] is that you'd have to change all existing job configurations to exclude "sig-windows". On the other hand, most jobs already skip "Feature:"-tagged tests.

**Comment by vyta** [comment]
@bclau yes, I missed updating the BUILD file in the e2e/windows directory. That should hopefully fix it

File: test/e2e/windows/README.md [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    * export KUBE_TEST_REPO_LIST=$(pwd)/repo_list.yaml
+
+1. Run only sig-windows tests:
+
+    ```./e2e.test --provider=local --ginkgo.noColor --ginkgo.focus=.*sig-windows*```
```
Comment (fejta) [inline test/e2e/windows/README.md]: Three ticks should have a line break after it, so

'''bash
./e2e.test --provider=local
'''

File: test/e2e/windows/density.go [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/test/e2e/framework"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+	"k8s.io/apimachinery/pkg/util/uuid"
+
+	. "github.com/onsi/ginkgo"
```
Comment (fejta) [inline test/e2e/windows/density.go]: Why are we using dot imports?

It seems deeply unlikely that ginkgo or gomega imports the windows package, so this is not an appropriate use of this functionality:

https://github.com/golang/go/wiki/CodeReviewComments#import-dot

**Review by fejta — COMMENTED** [review]
/approve

File: test/e2e/windows/density.go (hunk 10 above)
Comment (vyta) [reply test/e2e/windows/density.go]: According to ginkgo's docs, https://onsi.github.io/ginkgo/#avoiding-dot-imports, its _rare to avoid the dot imports... ginkgo bootstrap and generate default to dot imports. I can omit the dot, but seems better to not have to preface everything with ginkgo/gomega

**Comment by PatrickLang** [comment]
re: tags - we discussed this in SIG-Windows and I dropped a note on the groups thread with SIG-Architecture. We'll keep these tagged as [SIG-Windows] and include the tests in our existing prow job once this merges.

https://groups.google.com/d/msg/kubernetes-sig-architecture/MF50t_eTv3w/qWbJirXKFQAJ

**Comment by PatrickLang** [comment]
Few fixes in usage - adding the NodeDistro check means another commandline parameter is needed: https://github.com/PatrickLang/kubernetes/commit/254b3348e77836cd1e4e56ab81c27f229d030dea

File: test/e2e/windows/README.md (hunk 7 above)
Comment (PatrickLang) [reply test/e2e/windows/README.md]: This URL changed - I updated it in https://github.com/PatrickLang/kubernetes/commit/254b3348e77836cd1e4e56ab81c27f229d030dea

I think we should move it to the windows-testing repo, but not before this change. I think we need to get an automatic build running there first and the repo list should point to the output of that build.

**Comment by PatrickLang** [comment]
Other than the needed changes in the README (see above) looks good to me

**Comment by vyta** [comment]
@PatrickLang thanks! I'll add your commits

**Comment by PatrickLang** [comment]
/lgtm 
I manually tested on Windows Server 1803 & passed
